### PR TITLE
Created `onlySupportedEntryPoint(...)` modifier avoiding code redundancy

### LINKED ISSUE #157: [4337] Using `modifier` for checking `SUPPORTED_ENTRYPOINT`
## Description

The EntryPoint is checked in the validation and execution of user operation in 4337 which is done as below:

```
address entryPoint = _msgSender();
require(entryPoint == SUPPORTED_ENTRYPOINT, "Unsupported entry point");
```

The above code lines are duplicated in three functions (`validateUserOp(...)`, `executeUserOp(...)` and `executeUserOpWithErrorString(...)`) which can be abstracted away with a `modifier`.

## Proposed Solution

Using a modifier which does the same check and integrating it with the above-mentioned functions for better readable code.

### PR DESCRIPTION
**Pull request by remedcu**
Closes #157

This PR removes the redundant code to check if the entry point contract is actually supported or not.

### CODE REVIEW

**Comment by remedcu** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by mmv08 — COMMENTED** [review]
I don't know if it's worth checking this for a modifier used only thrice, but I'm curious about the impact on the code size? (we have a code size hardhat task)

**Comment by remedcu** [comment]
> I don't know if it's worth checking this for a modifier used only thrice, but I'm curious about the impact on the code size? (we have a code size hardhat task)

Before:
```
Safe4337Module 8100 bytes (limit is 24576)
```

After:
```
Safe4337Module 8135 bytes (limit is 24576)
```

**35 bytes higher.**

I checked by changing the `modifier` to an `internal` function. And then the code size was `7767 bytes`, **around 300+ bytes smaller**. Should I make the change?

**Comment by mmv08** [comment]
> I checked by changing the modifier to an internal function. And then the code size was 7767 bytes, around 300+ bytes smaller. Should I make the change?

I'd say so. You could also keep the modifier and call that internal function in the modifier.

**Comment by mmv08** [comment]
> > call that internal function in the modifier
> 
> The downside is slightly higher runtime cost. Personally, I would trade improving runtime cost for codesize here, because this contract is a “deploy once, use 1000000000 times” kind of contract.

Does this apply to the internal function + modifier or standalone internal function as well?

**Comment by remedcu** [comment]
Based on checking with the demo code, it does apply to both using an internal function and an internal + modifier.

Source Demo Code: https://pastebin.com/M2qdK91m

**Comment by mmv08** [comment]
> Based on checking with the demo code, it does apply to both using an internal function and an internal + modifier.
> 
> Source Demo Code: https://pastebin.com/M2qdK91m

Nice, thanks for the analysis

File: 4337/test/eip4337/EIP4337Module.spec.ts [hunk 1]
```diff
@@ -81,7 +81,7 @@ describe('Safe4337Module', () => {
         signature,
       })
 
-      await expect(safeModule.validateUserOp(userOp, ethers.ZeroHash, 0)).to.be.revertedWith('Unsupported execution function id')
+      await expect(entryPoint.executeUserOp(userOp, 0)).to.be.revertedWith('Unsupported execution function id')
```
Comment (mmv08) [inline 4337/test/eip4337/EIP4337Module.spec.ts:84]: Why was this changed?

Comment (remedcu) [reply 4337/test/eip4337/EIP4337Module.spec.ts:84]: As the modifier gets executed first, this will lead to getting the error `Unsupported entry point` instead of `Unsupported execution function id`.

File: 4337/contracts/Safe4337Module.sol [hunk 2]
```diff
@@ -63,7 +75,6 @@ contract Safe4337Module is IAccount, HandlerContext, CompatibilityFallbackHandle
         );
 
         address entryPoint = _msgSender();
```
Comment (nlordell) [inline 4337/contracts/Safe4337Module.sol:77]: ubernit: Can you move this down into the `if` block where it is used?

Comment (remedcu) [reply 4337/contracts/Safe4337Module.sol:77]: Actually, it is used before the `if` block for `validationData`.

**Review by nlordell — COMMENTED** [review]
Code changes look good in general, just have the same question as @mmv08

File: 4337/test/eip4337/EIP4337Module.spec.ts (hunk 1 above)
Comment (nlordell) [reply 4337/test/eip4337/EIP4337Module.spec.ts:84]: Another way to change this while keeping the test calling `validateUserOp` would be to call the `validateUserOp` function from the `entryPoint` contract.

I would slightly prefer that over calling `executeUserOp` which only implicitely calls `validateUserOp`

File: 4337/contracts/Safe4337Module.sol (hunk 2 above)
Comment (nlordell) [reply 4337/contracts/Safe4337Module.sol:77]: 🤦 oops, sorry.

File: 4337/test/eip4337/EIP4337Module.spec.ts (hunk 1 above)
Comment (remedcu) [reply 4337/test/eip4337/EIP4337Module.spec.ts:84]: Something like this will do the trick @nlordell 

```solidity=
const entryPointAddress = await ethers.getSigner(await entryPoint.getAddress())
const safeFromEntryPoint = safeModule.connect(entryPointAddress)
await expect(safeFromEntryPoint.validateUserOp.staticCall(userOp, ethers.ZeroHash, 0)).to.be.revertedWith('Unsupported execution function id')
```

**Review by mmv08 — APPROVED** [review]
👏
